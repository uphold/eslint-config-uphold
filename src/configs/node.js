/**
 * Module dependencies.
 *
 * @typedef {import('eslint').Linter.Config} LinterConfig
 */

import { defineConfig } from 'eslint/config';
import babelParser from '@babel/eslint-parser';
import globals from 'globals';
import upholdBaseConfig from './base.js';

/**
 * Configuration for bin and scripts files.
 *
 * @type {LinterConfig}
 */

const upholdBinScriptsConfig = {
  files: ['**/bin/**', '**/scripts/**'],
  name: 'uphold/scripts',
  rules: {
    'no-console': 'off',
    'node-plugin/no-process-exit': 'off'
  }
};

/**
 * Shared configuration preset for `node`.
 */

const upholdNode = defineConfig([
  {
    extends: [upholdBaseConfig, upholdBinScriptsConfig],
    languageOptions: {
      ecmaVersion: 2020,
      globals: {
        ...globals.jasmine,
        ...globals.jest,
        ...globals.mocha,
        ...globals.node
      },
      parser: babelParser,
      parserOptions: {
        requireConfigFile: false
      },
      sourceType: 'module'
    },
    name: 'uphold/node'
  }
]);

/**
 * Export the configuration.
 *
 * @see https://nodejs.org/docs/latest-v22.x/api/modules.html#loading-ecmascript-modules-using-require
 */

export { upholdNode as 'module.exports' };
export default upholdNode;
