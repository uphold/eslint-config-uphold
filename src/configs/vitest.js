/**
 * Module dependencies.
 */

import { isModuleAvailable, loadModule } from '../utils/load-module.js';

/**
 * Check if `@vitest/eslint-plugin` and its dependencies are available.
 */

const isTypeScriptAvailable = isModuleAvailable('typescript');
const isVitestPluginAvailable = isModuleAvailable('@vitest/eslint-plugin');
const isVitestAvailable = isModuleAvailable('vitest');

/**
 * Uphold Vitest ESLint config.
 * @type {import('@eslint/config-helpers').ConfigWithExtends}
 */
let upholdVitestConfig;

if (isVitestPluginAvailable && isTypeScriptAvailable && isVitestAvailable) {
  const { default: vitestPlugin } = await loadModule('@vitest/eslint-plugin');

  upholdVitestConfig = {
    extends: [vitestPlugin.configs.recommended],
    languageOptions: vitestPlugin.configs.env.languageOptions,
    name: 'uphold/vitest-plugin-config',
    rules: {
      'vitest/no-commented-out-tests': 'warn',
      'vitest/no-disabled-tests': 'warn',
      'vitest/no-focused-tests': 'error',
      'vitest/no-interpolation-in-snapshots': 'error',
      'vitest/no-mocks-import': 'error',
      'vitest/no-standalone-expect': 'error',
      'vitest/no-test-prefixes': 'error',
      'vitest/prefer-expect-resolves': 'warn',
      'vitest/prefer-to-have-length': 'warn'
    }
  };
} else if (isVitestAvailable) {
  if (!isTypeScriptAvailable) {
    console.warn('`typescript` is not installed, Vitest linting will be disabled');
  }

  if (!isVitestPluginAvailable) {
    console.warn('`@vitest/eslint-plugin` is not installed, Vitest linting will be disabled');
  }

  const globals = await loadModule('globals');

  upholdVitestConfig = {
    languageOptions: {
      globals: globals.vitest
    },
    name: 'uphold/vitest-globals'
  };
} else {
  console.error('Vitest is not installed, this config will be empty');

  upholdVitestConfig = {
    name: 'uphold/vitest-empty'
  };
}

/**
 * Export the configuration.
 */

export { upholdVitestConfig as 'module.exports' };
export default upholdVitestConfig;
