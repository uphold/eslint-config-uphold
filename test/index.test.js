/**
 * Module dependencies.
 */

import { ESLint } from 'eslint';
import { describe, expect, it } from 'vitest';
import { resolve } from 'node:path';
import baseConfig from '../src/index.js';

/**
 * Instances.
 */

const linter = new ESLint({ baseConfig, ignore: false });

/**
 * Test `eslint-config-uphold`.
 */

describe('eslint-config-uphold', () => {
  it('should not generate any violation for correct code', async () => {
    const source = resolve(import.meta.dirname, 'fixtures/correct.js');
    const [result] = await linter.lintFiles([source]);

    expect(result.messages).toHaveLength(0);
  });

  it('should generate violations for incorrect code', async () => {
    const source = resolve(import.meta.dirname, 'fixtures/incorrect.js');
    const [result] = await linter.lintFiles([source]);
    const rules = result.messages.map(violation => violation.ruleId);
    const violations = result.messages.map(({ column, line, ruleId, severity }) => ({
      column,
      line,
      ruleId,
      severity
    }));

    expect(violations).toMatchSnapshot();
    expect(Array.from(rules)).toEqual([
      'array-callback-return',
      'no-console',
      'consistent-this',
      'curly',
      'dot-notation',
      'id-match',
      'jsdoc/require-description-complete-sentence',
      'jsdoc/require-description-complete-sentence',
      'mocha/no-exclusive-tests',
      'mocha/no-identical-title',
      'mocha/no-nested-tests',
      'mocha/no-sibling-hooks',
      'no-new',
      'new-cap',
      'no-class-assign',
      'no-const-assign',
      'no-constant-condition',
      'no-dupe-class-members',
      'no-dupe-keys',
      'no-irregular-whitespace',
      'no-irregular-whitespace',
      'prettier/prettier',
      'no-irregular-whitespace',
      'no-irregular-whitespace',
      'no-labels',
      'no-labels',
      'prettier/prettier',
      'no-multi-str',
      'no-this-before-super',
      'no-underscore-dangle',
      'no-unused-vars',
      'prefer-destructuring',
      'prefer-destructuring',
      'prettier/prettier',
      'prettier/prettier',
      'promise/prefer-await-to-then',
      'uphold-plugin/explicit-sinon-use-fake-timers',
      'sort-destructure-keys/sort-destructure-keys',
      'sort-destructure-keys/sort-destructure-keys',
      'sort-destructure-keys/sort-destructure-keys',
      'sort-imports-requires/sort-imports',
      'sort-imports-requires/sort-requires',
      'sort-keys-fix/sort-keys-fix',
      'stylistic/padding-line-between-statements',
      'stylistic/padding-line-between-statements',
      'stylistic/spaced-comment',
      'sql-template/no-unsafe-query',
      'yoda'
    ]);
  });

  it('should not generate any violation for correct code inside bin & scripts folders', async () => {
    const binCorrect = resolve(import.meta.dirname, 'fixtures/bin/correct.js');
    const scriptsCorrect = resolve(import.meta.dirname, 'fixtures/scripts/correct.js');
    const [binResult, scriptResult] = await linter.lintFiles([binCorrect, scriptsCorrect]);

    expect(binResult.messages).toHaveLength(0);
    expect(scriptResult.messages).toHaveLength(0);
  });
});
