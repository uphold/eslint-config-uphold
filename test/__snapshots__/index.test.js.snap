// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`eslint-config-uphold > should generate violations for incorrect code 1`] = `
[
  {
    "column": 15,
    "line": 7,
    "ruleId": "array-callback-return",
    "severity": 2,
  },
  {
    "column": 1,
    "line": 10,
    "ruleId": "no-console",
    "severity": 1,
  },
  {
    "column": 7,
    "line": 13,
    "ruleId": "consistent-this",
    "severity": 2,
  },
  {
    "column": 12,
    "line": 20,
    "ruleId": "curly",
    "severity": 2,
  },
  {
    "column": 13,
    "line": 25,
    "ruleId": "dot-notation",
    "severity": 2,
  },
  {
    "column": 5,
    "line": 28,
    "ruleId": "id-match",
    "severity": 2,
  },
  {
    "column": 1,
    "line": 33,
    "ruleId": "jsdoc/require-description-complete-sentence",
    "severity": 2,
  },
  {
    "column": 1,
    "line": 33,
    "ruleId": "jsdoc/require-description-complete-sentence",
    "severity": 2,
  },
  {
    "column": 10,
    "line": 37,
    "ruleId": "mocha/no-exclusive-tests",
    "severity": 2,
  },
  {
    "column": 3,
    "line": 44,
    "ruleId": "mocha/no-identical-title",
    "severity": 2,
  },
  {
    "column": 3,
    "line": 49,
    "ruleId": "mocha/no-nested-tests",
    "severity": 2,
  },
  {
    "column": 3,
    "line": 55,
    "ruleId": "mocha/no-sibling-hooks",
    "severity": 2,
  },
  {
    "column": 1,
    "line": 61,
    "ruleId": "no-new",
    "severity": 2,
  },
  {
    "column": 5,
    "line": 61,
    "ruleId": "new-cap",
    "severity": 2,
  },
  {
    "column": 1,
    "line": 66,
    "ruleId": "no-class-assign",
    "severity": 2,
  },
  {
    "column": 1,
    "line": 73,
    "ruleId": "no-const-assign",
    "severity": 2,
  },
  {
    "column": 5,
    "line": 78,
    "ruleId": "no-constant-condition",
    "severity": 2,
  },
  {
    "column": 3,
    "line": 88,
    "ruleId": "no-dupe-class-members",
    "severity": 2,
  },
  {
    "column": 34,
    "line": 96,
    "ruleId": "no-dupe-keys",
    "severity": 2,
  },
  {
    "column": 28,
    "line": 101,
    "ruleId": "no-irregular-whitespace",
    "severity": 2,
  },
  {
    "column": 26,
    "line": 102,
    "ruleId": "no-irregular-whitespace",
    "severity": 2,
  },
  {
    "column": 45,
    "line": 102,
    "ruleId": "prettier/prettier",
    "severity": 2,
  },
  {
    "column": 45,
    "line": 102,
    "ruleId": "no-irregular-whitespace",
    "severity": 2,
  },
  {
    "column": 75,
    "line": 102,
    "ruleId": "no-irregular-whitespace",
    "severity": 2,
  },
  {
    "column": 1,
    "line": 105,
    "ruleId": "no-labels",
    "severity": 2,
  },
  {
    "column": 3,
    "line": 106,
    "ruleId": "no-labels",
    "severity": 2,
  },
  {
    "column": 19,
    "line": 110,
    "ruleId": "prettier/prettier",
    "severity": 2,
  },
  {
    "column": 20,
    "line": 110,
    "ruleId": "no-multi-str",
    "severity": 2,
  },
  {
    "column": 5,
    "line": 120,
    "ruleId": "no-this-before-super",
    "severity": 2,
  },
  {
    "column": 5,
    "line": 130,
    "ruleId": "no-underscore-dangle",
    "severity": 2,
  },
  {
    "column": 7,
    "line": 137,
    "ruleId": "no-unused-vars",
    "severity": 2,
  },
  {
    "column": 7,
    "line": 141,
    "ruleId": "prefer-destructuring",
    "severity": 2,
  },
  {
    "column": 7,
    "line": 142,
    "ruleId": "prefer-destructuring",
    "severity": 2,
  },
  {
    "column": 21,
    "line": 147,
    "ruleId": "prettier/prettier",
    "severity": 2,
  },
  {
    "column": 26,
    "line": 151,
    "ruleId": "prettier/prettier",
    "severity": 2,
  },
  {
    "column": 7,
    "line": 157,
    "ruleId": "promise/prefer-await-to-then",
    "severity": 2,
  },
  {
    "column": 1,
    "line": 166,
    "ruleId": "uphold-plugin/explicit-sinon-use-fake-timers",
    "severity": 2,
  },
  {
    "column": 15,
    "line": 177,
    "ruleId": "sort-destructure-keys/sort-destructure-keys",
    "severity": 2,
  },
  {
    "column": 16,
    "line": 178,
    "ruleId": "sort-destructure-keys/sort-destructure-keys",
    "severity": 2,
  },
  {
    "column": 22,
    "line": 179,
    "ruleId": "sort-destructure-keys/sort-destructure-keys",
    "severity": 2,
  },
  {
    "column": 1,
    "line": 190,
    "ruleId": "sort-imports-requires/sort-imports",
    "severity": 2,
  },
  {
    "column": 1,
    "line": 197,
    "ruleId": "sort-imports-requires/sort-requires",
    "severity": 2,
  },
  {
    "column": 3,
    "line": 206,
    "ruleId": "sort-keys-fix/sort-keys-fix",
    "severity": 2,
  },
  {
    "column": 1,
    "line": 213,
    "ruleId": "stylistic/padding-line-between-statements",
    "severity": 2,
  },
  {
    "column": 3,
    "line": 219,
    "ruleId": "stylistic/padding-line-between-statements",
    "severity": 2,
  },
  {
    "column": 1,
    "line": 225,
    "ruleId": "stylistic/spaced-comment",
    "severity": 2,
  },
  {
    "column": 10,
    "line": 233,
    "ruleId": "sql-template/no-unsafe-query",
    "severity": 2,
  },
  {
    "column": 5,
    "line": 238,
    "ruleId": "yoda",
    "severity": 2,
  },
]
`;
